### PR TITLE
Lh/element optimizations

### PR DESCRIPTION
**Pull request by LHerskind**
A collection of gas optimizations related to the Element Finance bridge and the use of the heap. Mostly caching values instead of fetching from storage multiple times etc. 

Saves ~4K gas on the avg `convert` and ~2K on `finalise`.

### CODE REVIEW

File: src/bridges/element/ElementBridge.sol [hunk 1]
```diff
@@ -469,47 +475,57 @@ contract ElementBridge is IDefiBridge {
...
         // store the tranche that underpins our interaction, the expiry and the number of received tokens against the nonce
         Interaction storage newInteraction = interactions[convertArgs.interactionNonce];
-        newInteraction.expiry = trancheExpiry;
-        newInteraction.failed = false;
-        newInteraction.finalised = false;
         newInteraction.quantityPT = principalTokensAmount;
-        newInteraction.trancheAddress = pool.trancheAddress;
+        newInteraction.trancheAddress = trancheAddress;
+        newInteraction.expiry = trancheExpiry;
+        //newInteraction.failed = false;
```
Comment (PhilWindle) [inline src/bridges/element/ElementBridge.sol:500]: We can remove these commented out lines

**Review by PhilWindle — CHANGES_REQUESTED** [review]

File: src/bridges/element/ElementBridge.sol [hunk 2]
```diff
@@ -735,30 +755,37 @@ contract ElementBridge is IDefiBridge {
...
+            nonces[index] = nonces[noncesLength - 1];
         }
         nonces.pop();
 
         // if there are no more nonces left for this expiry then remove it from the heap
-        if (nonces.length == 0) {
-            heap.remove(interaction.expiry);
-            delete expiryToNonce[interaction.expiry];
-            return (true);
+        if (noncesLength == 0) {
```
[the remark this answers is not in the document]
Comment (LHerskind) [reply src/bridges/element/ElementBridge.sol]: Good catch. Fixed in https://github.com/AztecProtocol/aztec-connect-bridges/pull/67/commits/5fd7fcbab2747b62ec6bee7b8eb2fdb767ec14f7

File: src/bridges/element/ElementBridge.sol [hunk 3]
```diff
@@ -469,47 +475,57 @@ contract ElementBridge is IDefiBridge {
...
         }
-        ITranche tranche = ITranche(pool.trancheAddress);
+        ITranche tranche = ITranche(trancheAddress);
         uint64 trancheExpiry = uint64(tranche.unlockTimestamp());
         if (block.timestamp >= trancheExpiry) {
             revert TRANCHE_ALREADY_EXPIRED();
         }
-        
+
         // approve the transfer of tokens to the balancer address
```
[the remark this answers is not in the document]
Comment (LHerskind) [reply src/bridges/element/ElementBridge.sol]: Done in https://github.com/AztecProtocol/aztec-connect-bridges/pull/67/commits/a26a63614910d59176ff69ff9ce4c8bcd8704388

File: src/bridges/element/ElementBridge.sol (hunk 1 above)
Comment (LHerskind) [reply src/bridges/element/ElementBridge.sol]: Updated in https://github.com/AztecProtocol/aztec-connect-bridges/pull/67/commits/172d2833533797e11d426f7f4e27eee119bb2046

File: src/bridges/element/ElementBridge.sol [hunk 4]
```diff
@@ -525,29 +541,13 @@ contract ElementBridge is IDefiBridge {
...
 
-        uint256 trancheTokenQuantityBefore = ERC20(pool.trancheAddress).balanceOf(address(this));
+        ERC20(inputAsset).approve(balancerAddress, inputQuantity);
         quantityReceived = IVault(balancerAddress).swap(
             singleSwap,
             fundManagement,
             inputQuantity, // we won't accept less than 1 output token per input token
             block.timestamp
         );
-
```
[the remark this answers is not in the document]
Comment (LHerskind) [reply src/bridges/element/ElementBridge.sol]: Readded in https://github.com/AztecProtocol/aztec-connect-bridges/pull/67/commits/172d2833533797e11d426f7f4e27eee119bb2046

**Review by PhilWindle — APPROVED** [review]

**Pull request merged**